prow/googlecloudbuild/client/client.go

```diff
@@ -0,0 +1,182 @@
+)
+
+const (
+	// GCB automatically assigns random build ids to a new build, so it's not
+	// possible to associate a GCB build with a prow job by id without storing
+	// mapping somewhere else. Utilizing GCB tags for achieving this.
+
+	// ProwLabelSeparater formats labels key:val pairs into `{key} ::: {val}`
+	ProwLabelSeparater = " ::: "
+)
+
+// Operator is the interace that's highly recommended to be used by any package
+// that imports current package. Unit test can be carried out with a mock
+// implemented under fake of current package
+type Operator interface {
+	GetBuild(ctx context.Context, project, id string) (*cloudbuildpb.Build, error)
+	ListBuildsByTag(ctx context.Context, project string, tags []string) ([]*cloudbuildpb.Build, error)
+	CreateBuild(ctx context.Context, project string, bld *cloudbuildpb.Build) (*cloudbuildpb.Build, error)
+	CancelBuild(ctx context.Context, project, id string) (*cloudbuildpb.Build, error)
+}
+
+// ProwLabel formats labels key:val pairs into `{key} ::: {val}`
+func ProwLabel(key, val string) string {
+	return key + ProwLabelSeparater + val
+}
+
+// LabelFromTag trims label into key:val pairs. returns empty string as value if
+// the label is not formatted as prow label format
+func LabelFromTag(tag string) (key, val string) {
+	parts := strings.SplitN(tag, ProwLabelSeparater, 2)
+	if len(parts) < 2 {
+		return tag, ""
+	}
+	return parts[0], parts[1]
+}
+
+// GetProwLabels gets labels from cloud build struct, simulating k8s pods labels format
+// of map[string]string
+func GetProwLabels(bld *cloudbuildpb.Build) map[string]string {
+	res := make(map[string]string)
+	for _, tag := range bld.Tags {
+		key, val := LabelFromTag(tag)
+		if val != "" {
+			res[key] = val
+		}
+	}
+	return res
+}
+
+// Client wraps native cloudbuild client
+type Client struct {
+	interactor *cloudbuild.Client
+}
+
+// NewClient creates a new Client, with optinal credentialFile
+func NewClient(ctx context.Context, credentialFile string) (*Client, error) {
+	var opts []option.ClientOption
+	// Authenticating with key file if it's provided
+	if len(credentialFile) > 0 {
+		if _, err := os.Stat(credentialFile); err != nil {
+			return nil, err
+		}
+		opts = append(opts, option.WithCredentialsFile(credentialFile))
+	}
+
+	cbClient, err := cloudbuild.NewClient(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{interactor: cbClient}, nil
+}
+
+// GetBuild gets build by it's id
+func (c *Client) GetBuild(ctx context.Context, project, id string) (*cloudbuildpb.Build, error) {
+	return c.interactor.GetBuild(ctx, &cloudbuildpb.GetBuildRequest{
+		ProjectId: project,
+		Id:        id,
+	})
+}
+
+// ListBuildsByTag lists builds by tags
+func (c *Client) ListBuildsByTag(ctx context.Context, project string, tags []string) ([]*cloudbuildpb.Build, error) {
+	const pageSize = 50
+	var res []*cloudbuildpb.Build
+	var tagsFilters []string
+	for _, tag := range tags {
+		tagsFilters = append(tagsFilters, "tags="+tag)
```

Comment:
So given tags ["a", "b", "c"], tagsFilters will look like ["tags=a", "tags=b", "tags=c"] ? Is this right? (Given the plural of `tags` I was expecting something like `tags=a,b,c`).

EDIT: I see that you do a ` AND ` strings.Join below, so it makes more sense now (although on first glance it still looks a bit awkward).

Should we be doing any empty string checks here? E.g., `tags=` by itself will probably break the Filter, right? E.g., `tags= AND tags=foo` will probably not work?

Reply:
good point, yeah "tags=" would break. However I'm a bit split here, on the one hand we should guard against empty string, on the other hand, I also feel like this is caller's responsibility